### PR TITLE
=htp #20052 headerValueByType now works with custom headers

### LINKED ISSUE #20052: headerValueByType fails to match
At least when using a `ModeledCustomHeader`, the directive `headerValueByType` fails to match. To reproduce the error, take a look at the following test case from akka-sse and replace the working `headerValueByName` with the commented `headerValueByType`: https://github.com/hseeberger/akka-sse/blob/master/akka-sse/src/test/scala/de/heikoseeberger/akkasse/headers/LastEventIdSpec.scala#L45

### PR DESCRIPTION
**Pull request by ktoso**
Resolves #20052

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala [hunk 1]
```diff
@@ -87,6 +87,7 @@ abstract class ModeledCustomHeaderCompanion[H <: ModeledCustomHeader[H]] {
     case _               ⇒ None
   }
 
+  final implicit val implicitlyLocatableCompanion: ModeledCustomHeaderCompanion[H] = this
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala]: This allows us to implicitly find the companion, if we know the type of a header (as given `T`, implicit search scope includes the `T` object, in which we find this)

File: akka-http-core/src/test/scala/akka/http/javadsl/HttpExtensionApiSpec.scala [hunk 2]
```diff
@@ -380,7 +380,6 @@ class HttpExtensionApiSpec extends WordSpec with Matchers with BeforeAndAfterAll
 
     "create an outgoing connection (with 6 parameters)" in {
       val (host, port, binding) = runServer()
-      println("host = " + host)
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/javadsl/HttpExtensionApiSpec.scala]: randomly found and fixed

**Comment by ktoso** [comment]
Any opinions here?

**Comment by ktoso** [comment]
Anyone happy to review implicit magics? :wink: 

@akka/akka-team

**Comment by patriknw** [comment]
LGTM, I guess

**Comment by ktoso** [comment]
I promise to write a blog post explaining it :wink:

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala (hunk 1 above)
Comment (rkuhn) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala]: sneaky

File: akka-http-core/src/test/scala/akka/http/javadsl/HttpExtensionApiSpec.scala (hunk 2 above)
Comment (rkuhn) [inline akka-http-core/src/test/scala/akka/http/javadsl/HttpExtensionApiSpec.scala]: :+1:

File: akka-http/src/main/scala/akka/http/javadsl/server/HttpService.scala [hunk 3]
```diff
@@ -7,7 +7,6 @@ package akka.http.javadsl.server
 import akka.actor.ActorSystem
 import akka.http.scaladsl.{ server, Http }
 import akka.http.scaladsl.Http.ServerBinding
-import akka.http.scaladsl.server.RouteResult
```
Comment (rkuhn) [inline akka-http/src/main/scala/akka/http/javadsl/server/HttpService.scala]: might be better to just `import akka.http.scaladsl` to make usage really obvious

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala [hunk 4]
```diff
@@ -5,7 +5,13 @@
 package akka.http.scaladsl.server
 package directives
 
+import akka.http.javadsl.model.headers.CustomHeader
+import akka.http.scaladsl.model.headers.{ModeledCustomHeaderCompanion, ModeledCustomHeader, RawHeader}
+
+import scala.annotation.implicitNotFound
```
Comment (rkuhn) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala]: I don’t see this used, maybe it should?

**Comment by rkuhn** [comment]
LGTM

**Pull request merged**